2996_reflection/reflection.md

```diff
@@ -999,6 +1000,87 @@ Thus, the only instantiations of `TU_Ticket::Helper` occur because of the call t
 
 [On Compiler Explorer](https://godbolt.org/z/1vEjW4sTr).
 
+## Emulating typeful reflection
+Although we believe a single opaque `std::meta::info` type to be the best and most scalable foundation for reflection, we acknowledge the desire expressed by SG7 for future support for "typeful reflection". The following demonstrates one possible means of assembling a typeful reflection library, in which different classes of reflections are represented by distinct types, on top of the facilities proposed here.
+
+::: bq
+```cpp
+// Represents a 'std::meta::info' constrained by a predicate.
+template <std::meta::info Pred>
+  requires (type_of(^([:Pred:](^int))) == ^bool)
+struct metatype {
+  std::meta::info value;
+
+  // Construction is ill-formed unless predicate is satisfied.
+  consteval metatype(std::meta::info r) : value(r) {
+    if (![:Pred:](r))
+      throw "Reflection is not a member of this metatype";
+  }
+
+  // Cast to 'std::meta::info' allows values of this type to be spliced.
+  consteval operator std::meta::info() const { return value; }
+
+  static consteval bool check(std::meta::info r) { return [:Pred:](r); }
+};
+
+// Type representing a "failure to match" any known metatypes.
+struct unmatched {
+  consteval unmatched(std::meta::info) {}
+  static consteval bool check(std::meta::info) { return true; }
+};
+
+// Returns the given reflection "enriched" with a more descriptive type.
+template <typename... Choices>
+consteval std::meta::info enrich(std::meta::info r) {
+  std::array ctors = {members_of(^Choices, std::meta::is_constructor)[0]...,
+                      members_of(^unmatched, std::meta::is_constructor)[0]};
```

Comment:
Why only taking the first constructor? I guess what you need is the constructor that takes a `std::meta::info`, but this relies on that one being ahead of the copy and move constructors?

Reply:
Well I control the class, and the Standard guarantees that implicitly-declared special member functions are considered to be declared at the closing `}` of the class definition:
https://eel.is/c++draft/class#special

> An implicitly-declared special member function is declared at the closing } of the class-specifier.

And `members_of` guarantees iteration in declaration order, so for the purpose of this example, this seems safe to me.